proposals/0000-autolinking-local-modules.md

```diff
@@ -0,0 +1,83 @@
+---
+title: Autolinking local modules
+author:
+- Szymon Rybczak
+- Satyajit Sahoo
+date: 21-08-2023
+---
+
+# RFC0000: Autolinking local modules
+
+## Summary
+
+This RFC proposes a new feature to [React Native CLI](https://github.com/react-native-community/cli/) where we can create a folder containing local native modules and views and have them linked automatically. Proposed solution will simplify the space of creating local native modules and views and using them later in the app.
+
+## Motivation
+
+Currently, when we have local native modules, we have to create them under `android` and `ios` folders and write the code for registering those packages manually. This poses a few problems:
+
+1. The new modules become part of the `android` and `ios` source code—which means upgrading the project to a newer React Native version is harder as it also needs to consider those changes,
+2. The module boilerplate is written manually, that's the place where users can make mistakes and it takes a lot of time to write one,
+3. The code for registering the modules is additional boilerplate which is not required for third party packages (because they're automatically linked),
+4. The workflow for creating new architecture modules without creating packages is not streamlined,
+5. It's challenging to open-source the local module, as it requires creating a dedicated library setup that adds extra friction to the process.
+
+## Detailed design
+
+To automate the process of creating native modules or views we need to implement a few things.
+First we need to introduce a new feature to the `create-react-native-library` package to create local modules. `create-react-native-library` will detect if the command has been run under React Native project if yes, it will create a local module under `local_modules` directory. CRNL will contain templates for creating native modules and views, also templates for turbo modules and fabric views.
```

Comment:
Q: What's the difference from this approach from just using `create-react-native-library` inside the `packages/` folder and `yarn add`-ing it?

Wouldn't it make it easier to open-source actually, as you're creating a proper NPM package?

Reply:
In other word: do we need this? What problem is this solving that CRNL isn't solving already?

Reply:
@cortinico right now, the code generated by `create-react-native-library` includes an `example` folder which isn't necessary for local modules, additional dependencies in the `package.json`, build scripts for publishing etc. that need to be maintained separately. the idea with `local_modules` is that we remove the additional unnecessary stuff to give a smaller amount of stuff to maintain. It's harder to open source than a regular lib, but easier to mainatin.

`yarn add`ing an repo like this also has other issues. it doesn't behave exactly like installing a package from `npm`, it could result in duplicate dependencies etc. due to the additional `node_modules` which could cause issues.